package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,943 @@
+#
+# Neighbor Search Stuff
+#
+########################################################################################################################
+cdef class NSResults(object):
+    cdef readonly real cutoff
+    cdef ns_int npairs
+    cdef bint debug
+
+    cdef real[:, ::1] coords # shape: size, DIM
+    cdef ns_int[:] search_ids
+
+    cdef ns_int allocation_size
+    cdef ipair *pairs # shape: pair_allocation
+    cdef real *pair_distances2 # shape: pair_allocation
+
+    cdef list indices_buffer
+    cdef list coordinates_buffer
+    cdef list distances_buffer
+    cdef np.ndarray pairs_buffer
+    cdef np.ndarray pair_distances_buffer
+    cdef np.ndarray pair_coordinates_buffer
+
+    def __init__(self, real cutoff, real[:, ::1]coords, ns_int[:] search_ids, debug=False):
+
+        self.debug = debug
+        self.cutoff = cutoff
+        self.coords = coords
+        self.search_ids = search_ids
+
+        # Preallocate memory
+        self.allocation_size = search_ids.shape[0] + 1
+        self.pairs = <ipair *> PyMem_Malloc(sizeof(ipair) * self.allocation_size)
+        if not self.pairs:
+            MemoryError("Could not allocate memory for NSResults.pairs ({} bits requested)".format(sizeof(ipair) * self.allocation_size))
+        self.pair_distances2 = <real *> PyMem_Malloc(sizeof(real) * self.allocation_size)
+        if not self.pair_distances2:
+            raise MemoryError("Could not allocate memory for NSResults.pair_distances2 ({} bits requested)".format(sizeof(real) * self.allocation_size))
+        self.npairs = 0
+
+        # Buffer
+        self.indices_buffer = None
+        self.coordinates_buffer = None
+        self.distances_buffer = None
+        self.pairs_buffer = None
+        self.pair_coordinates_buffer = None
+
+    def __dealloc__(self):
+        PyMem_Free(self.pairs)
+        PyMem_Free(self.pair_distances2)
+
+    cdef int add_neighbors(self, ns_int beadid_i, ns_int beadid_j, real distance2) nogil:
+        # Important: If this function returns 0, it means that memory allocation failed
+
+        # Reallocate memory if needed
+        if self.npairs >= self.allocation_size:
+            # We need to reallocate memory
+            if self.resize(self.allocation_size + <ns_int> (self.allocation_size * 0.5 + 1)) == 0:
+                return 0
+
+        # Actually store pair and distance squared
+        if beadid_i < beadid_j:
+            self.pairs[self.npairs][0] = beadid_i
+            self.pairs[self.npairs][1] = beadid_j
+        else:
+            self.pairs[self.npairs][1] = beadid_i
+            self.pairs[self.npairs][0] = beadid_j
+        self.pair_distances2[self.npairs] = distance2
+        self.npairs += 1
+
+        return self.npairs
+
+    cdef int resize(self, ns_int new_size) nogil:
+        # Important: If this function returns 0, it means that memory allocation failed

```

Comment:
This is counterintuitive, most other things returning 0 is no errors

Reply:
we could use some constants in the beginngin like `OK` and `ERROR`.